lib/shrine/storage/google_cloud_storage.rb

```diff
@@ -1,129 +1,111 @@
 require "shrine"
-require "googleauth"
-require "google/apis/storage_v1"
+require "google/cloud/storage"
 
 class Shrine
   module Storage
     class GoogleCloudStorage
       attr_reader :bucket, :prefix, :host
 
-      def initialize(bucket:, prefix: nil, host: nil, default_acl: nil, object_options: {})
+      # Initialize a Shrine::Storage for GCS allowing for auto-discovery of the Google::Cloud::Storage client.
+      # @param [String] project Provide if not using auto discovery
+      # @see http://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud-storage/v1.6.0/guides/authentication#environmentvariables for information on discovery
+      def initialize(project: nil, bucket:, prefix: nil, host: nil, default_acl: nil, object_options: {})
+        @project = project
         @bucket = bucket
         @prefix = prefix
         @host = host
         @default_acl = default_acl
         @object_options = object_options
       end
 
+      # If the file is an UploadFile from GCS, issues a copy command, otherwise it uploads a file.
+      # @param [IO] io - io like object
+      # @param [String] id - location
       def upload(io, id, shrine_metadata: {}, **_options)
-        # uploads `io` to the location `id`
-
-        object = Google::Apis::StorageV1::Object.new @object_options.merge(bucket: @bucket, name: object_name(id))
-
         if copyable?(io)
-          storage_api.copy_object(
-            io.storage.bucket,
-            io.storage.object_name(io.id),
-            @bucket,
-            object_name(id),
-            object,
-            destination_predefined_acl: @default_acl,
-          )
+          existing_file = get_bucket(io.storage.bucket).file(io.storage.object_name(io.id))
+          file = existing_file.copy(
+              @bucket, # dest_bucket_or_path - the bucket to copy the file to
+              object_name(id), # dest_path - the path to copy the file to in the given bucket
+              acl: @default_acl
+          ) do |f|
+            # update the additional options
+            @object_options.merge(_options).each_pair do |key, value|
```

Comment:
Rename to `options`. The `_` is a convention to indicate the argument is not used.